lib/pkcs7.js

```diff
@@ -537,6 +548,142 @@ p7.createSignedData = function() {
     // add signer info
     msg.signerInfos = _signersToAsn1(msg.signers);
   }
+
+  function verifySignerInfos(mds, caStore, options) {
+    var content;
+    var rval = true;
+    var svEvent = options.onSignatureVerificationComplete;
+
+    if(msg.contentInfo !== null && msg.contentInfo.value[1]) {
+      // Note: ContentInfo is a SEQUENCE with 2 values, second value is
+      // the content field and is optional for a ContentInfo but required here
+      // since signers are present
+      // get ContentInfo content
+      content = msg.contentInfo.value[1];
+      // skip [0] EXPLICIT content wrapper
+      content = content.value[0];
+    } else if('content' in msg) {
+      // signature was likely made in detached mode
+      // caller must set p7.content before attempting to verify
+      if(msg.content instanceof forge.util.ByteBuffer) {
+        content = msg.content.bytes();
+      } else if(typeof msg.content === 'string') {
+        content = forge.util.encodeUtf8(msg.content);
+      }
+      content = asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false, content);
+    }
+
+    if(!content) {
+      svEvent && svEvent(new Error('Could not verify PKCS#7 message; there is no content to verify.'), null);
+      return false;
+    }
+
+    if(msg.signers.length === 0) {
+      svEvent && svEvent(new Error('There are no signatures to verify.'), null);
+      return false;
+    }
+
+    // get ContentInfo content type
+    var contentType = asn1.derToOid(msg.contentInfo.value[0].value);
+
+    // serialize content
+    var bytes = asn1.toDer(content);
+
+    // skip identifier and length per RFC 2315 9.3
+    // skip identifier (1 byte)
+    bytes.getByte();
+    // read and discard length bytes
+    asn1.getBerValueLength(bytes);
+    bytes = bytes.getBytes();
```

Comment:
Does any deeper validation need to happen here?  Or if the stream is bogus will the digests just not match so that's fine?